Improve version resolution with osxcross-macports

ProjectSynchro opened the pull request:
I rejigged the package resolution logic to allow for older versions to be downloaded when the latest advertised one is not available for the target being requested.

When required and a fat binary is all that is available which contains the arch in question (e.g. a universal build contains arm64 but there is no arm64 package available) we will grab that.

Due to the change for macOS 11+ to use the arm64 packages by default, I added a flag to download x86_64 versions of packages so people building for x86 macs still can.

macports currently doesn't build arm64e universal packages, so when searching for newer darwin universal binaries it will fail, I removed arm64e from the list of arches.


Let me know if any changes look bad or if you'd like a change made. 

This makes the wrapper a bit more resilient for versions of Darwin where builds may be missing for a package version.

tpoechtrager commented:
Thanks!

`sort -V` is not available on *BSD. Let's make sure, we use `gsort` there. Other than that it looks good.

    if uname | grep -qi bsd; then
      SORT="gsort"
    else
      SORT="sort"
    fi
    
    require $SORT
    require "openssl"
    require "bzip2"
    require "tar"

Then use `$SORT` everywhere.

As a bonus we could also allow `-x86-64` and `--x86-64` as aliases for `-x86_64` and `--x86-64`.

ProjectSynchro commented:
> Thanks!
> 
> `sort -V` is not available on *BSD. Let's make sure, we use `gsort` there. Other than that it looks good.
> 
> ```
> if uname | grep -qi bsd; then
>   SORT="gsort"
> else
>   SORT="sort"
> fi
> 
> require $SORT
> require "openssl"
> require "bzip2"
> require "tar"
> ```
> 
> Then use `$SORT` everywhere.
> 
> As a bonus we could also allow `-x86-64` and `--x86-64` as aliases for `-x86_64` and `--x86-64`.

Sounds good! I'll adjust the implementation when I get a chance tomorrow.

ProjectSynchro commented:
Got to it sooner :smile: 

Should have everything you requested (Can squash if you want the history cleaner)

tpoechtrager commented:
The `sort` calls must be replaced with `$SORT`. Other than that, it looks good. Squashing it into a single commit would be great. It makes cherry-picking for the LLVM branch easier...

ProjectSynchro commented:
Whoops, forgot a replacement. Will do

ProjectSynchro committed:
Improve version resolution with osxcross-macports

I rejigged the package resolution logic to allow for older versions to be downloaded when the latest advertised one is not available for the target being requested.

When required and a fat binary is all that is available which contains the arch in question (e.g. a universal build contains arm64 but there is no arm64 package available) we will grab that.

Due to the change for macOS 11+ to use the arm64 packages by default, I added a flag to download x86_64 versions of packages so people building for x86 macs still can.

macports currently doesn't build arm64e universal packages, so when searching for newer darwin universal binaries it will fail, I removed arm64e from the list of arches.

This makes the wrapper a bit more resilient for versions of Darwin where builds may be missing for a package version.

ProjectSynchro commented:
Should be done now :)

tpoechtrager commented:
👍Thanks.
